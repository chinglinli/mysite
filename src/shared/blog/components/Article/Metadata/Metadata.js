import React from 'react'
import PropTypes from 'prop-types'
import ImmutablePropTypes from 'react-immutable-proptypes'
import NormalLink from '../../generic/NormalLink'
import styled from 'styled-components'

import dateformat from 'dateformat'

const ItemGroup = styled.ul`
  font-size: 0.9rem;
  overflow: hidden;
  text-overflow: ellipsis;
  white-space: nowrap;
`

const Item = styled.li`
  display: inline;
  &:not(:first-child) {
    margin-left: 8px;
  }
`

const Categories = styled.ul`
  display: inline;
`

const Category = styled.li`
  margin: 0;
  display: inline;
  &:not(:first-child):before {
    content: "/";
    margin: 0 3px;
  }
`

export default class Metadata extends React.PureComponent {
  static PropTypes = {
    categories: ImmutablePropTypes.listOf(ImmutablePropTypes.contains({
      slug: PropTypes.string.isRequired,
      name: PropTypes.string.isRequired
    })).isRequired,
    date: PropTypes.instanceOf(Date).isRequired,
    modifiedDate: PropTypes.instanceOf(Date)
  }

  renderCategoryItem = (categories) => {
    return (
      <Item>分類：
        <Categories>
          {categories.map(category => (
            <Category key={category.get('slug')}>
              <NormalLink to={`/articles/category/${category.get('slug')}/`}>{category.get('name')}</NormalLink>
            </Category>
          ))}
        </Categories>
      </Item>
    )
  }

  renderReleaseDateItem = (date) => {
    return (
      <Item>發表時間：{dateformat(date, 'yyyy/mm/dd')}</Item>
    )
  }

  renderModifiedDateItem = (modifiedDate) => {
    return (
      <Item>最後更新：{dateformat(modifiedDate, 'yyyy/mm/dd')}</Item>
    )
  }

  render () {
    return (
      <ItemGroup>
        {this.renderCategoryItem(this.props.categories)}
        {this.renderReleaseDateItem(this.props.date)}
        {this.props.modifiedDate && this.renderModifiedDateItem(this.props.modifiedDate)}
      </ItemGroup>
    )
  }
}

