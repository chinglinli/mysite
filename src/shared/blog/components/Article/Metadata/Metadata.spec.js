import Immutable from 'immutable'
import React from 'react'
import renderer from 'react-test-renderer'
import 'jest-styled-components'

import Metadata from './Metadata'

test('It should render Metadata component correctly', () => {
  const categories = Immutable.fromJS([{
    slug: 'slug',
    name: 'name'
  },
  {
    slug: 'slug2',
    name: 'name2'
  }])
  const date = new Date(2017, 6, 9)

  const component = renderer.create(
    <Metadata categories={categories} date={date} />
  )
  let tree = component.toJSON()
  expect(tree).toMatchSnapshot()
})

test('It should render Metadata component correctly when target article has been modified', () => {
  const categories = Immutable.fromJS([{
    slug: 'slug',
    name: 'name'
  }])
  const date = new Date(2017, 6, 9)
  const modifiedDate = new Date(2017, 10, 9)

  const component = renderer.create(
    <Metadata categories={categories} date={date} modifiedDate={modifiedDate} />
  )
  let tree = component.toJSON()
  expect(tree).toMatchSnapshot()
})
