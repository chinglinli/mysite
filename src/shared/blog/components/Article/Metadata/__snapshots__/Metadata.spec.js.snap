// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`It should render Metadata component correctly 1`] = `
<ul
  className="Metadata__ItemGroup-dlupfj-0 llqsil"
>
  <li
    className="Metadata__Item-dlupfj-1 fpRYfi"
  >
    分類：
    <ul
      className="Metadata__Categories-dlupfj-2 CNRul"
    >
      <li
        className="Metadata__Category-dlupfj-3 cJIpQm"
      >
        <a
          className="NormalLink-s877q9z-0 coaBWT"
          onClick={[Function]}
          style={Object {}}
        >
          name
        </a>
      </li>
      <li
        className="Metadata__Category-dlupfj-3 cJIpQm"
      >
        <a
          className="NormalLink-s877q9z-0 coaBWT"
          onClick={[Function]}
          style={Object {}}
        >
          name2
        </a>
      </li>
    </ul>
  </li>
  <li
    className="Metadata__Item-dlupfj-1 fpRYfi"
  >
    發表時間：
    2017/07/09
  </li>
</ul>
`;

exports[`It should render Metadata component correctly when target article has been modified 1`] = `
<ul
  className="Metadata__ItemGroup-dlupfj-0 llqsil"
>
  <li
    className="Metadata__Item-dlupfj-1 fpRYfi"
  >
    分類：
    <ul
      className="Metadata__Categories-dlupfj-2 CNRul"
    >
      <li
        className="Metadata__Category-dlupfj-3 cJIpQm"
      >
        <a
          className="NormalLink-s877q9z-0 coaBWT"
          onClick={[Function]}
          style={Object {}}
        >
          name
        </a>
      </li>
    </ul>
  </li>
  <li
    className="Metadata__Item-dlupfj-1 fpRYfi"
  >
    發表時間：
    2017/07/09
  </li>
  <li
    className="Metadata__Item-dlupfj-1 fpRYfi"
  >
    最後更新：
    2017/11/09
  </li>
</ul>
`;
